plugins/permission-backend/src/service/router.ts

```diff
@@ -93,28 +102,51 @@ const evaluatePermissionRequestBatchSchema: z.ZodSchema<EvaluatePermissionReques
  */
 export interface RouterOptions {
   logger: Logger;
-  discovery: PluginEndpointDiscovery;
+  discovery: DiscoveryService;
   policy: PermissionPolicy;
-  identity: IdentityApi;
+  identity?: IdentityApi;
   config: Config;
+  auth?: AuthService;
+  httpAuth?: HttpAuthService;
+  userInfo?: UserInfoService;
 }
 
 const handleRequest = async (
   requests: IdentifiedPermissionMessage<EvaluatePermissionRequest>[],
-  user: BackstageIdentityResponse | undefined,
   policy: PermissionPolicy,
   permissionIntegrationClient: PermissionIntegrationClient,
-  authHeader?: string,
+  credentials: BackstageCredentials<
+    BackstageNonePrincipal | BackstageUserPrincipal
+  >,
+  auth: AuthService,
+  userInfo: UserInfoService,
 ): Promise<IdentifiedPermissionMessage<EvaluatePermissionResponse>[]> => {
   const applyConditionsLoaderFor = memoize((pluginId: string) => {
     return new DataLoader<
       ApplyConditionsRequestEntry,
       ApplyConditionsResponseEntry
     >(batch =>
-      permissionIntegrationClient.applyConditions(pluginId, batch, authHeader),
+      permissionIntegrationClient.applyConditions(pluginId, credentials, batch),
     );
   });
 
+  let user: BackstageIdentityResponse | undefined;
+  if (auth.isPrincipal(credentials, 'user')) {
+    const { ownershipEntityRefs } = await userInfo.getUserInfo(credentials);
+    const { token } = await auth.getPluginRequestToken({
+      onBehalfOf: credentials,
+      targetPluginId: 'catalog', // TODO: unknown at this point
```

Comment:
?

Reply:
Yeah this is a quirk. It happens in at least one more place too in a different plugin.

The new system is based around credentials objects wrapping principals (for example a user). When you want to make an outbound call to a plugin, you need to request a token from the auth service, and you need to tell it what the target plugin is. The tokens are unique for each source-target plugin pair so they can't be leaked and used to call other services as well. You cannot pass incoming tokens straight through to outgoing calls either. Therefore, in legacy interfaces that ONLY accept a single static token, we are stuck until we can refactor the interface such that it accepts a credentials object instead of a token.

In this case, `PermissionPolicy.handle` takes a `BackstageIdentityResponse` which includes a token, and that is a bit unfortunate and might warrant a refactor down the line. That'll be a pretty breaking change though, as far as I can tell.

Reply:
slight correction: the token will only work towards the catalog **if** migrated to the new backend system **and** using our new auth service implementation that uses plugin-specific tokens, which we have not built yet and don't yet know if we will ship in 1.24.